aslam_offline_calibration/kalibr2/include/kalibr2/CalibrationTools.hpp

```diff
@@ -8,111 +8,333 @@
+ * @param[in] target The shared definition of the calibration target.
+ * @return The optimized 6-DOF transformation from the left camera to the right camera ($T_{camH\_camL}$).
+ * @throws std::runtime_error if the number of observations for the two cameras do not match, or if
+ * the optimization's linear solver fails.
+ */
+template <typename CameraLT, typename CameraHT>
+sm::kinematics::Transformation CalibrateStereoPair(
+    boost::shared_ptr<aslam::cameras::CameraGeometryBase> geometry_camera_L,
+    boost::shared_ptr<aslam::cameras::CameraGeometryBase> geometry_camera_H,
+    const std::vector<std::optional<aslam::cameras::GridCalibrationTargetObservation>>& observations_camera_L,
+    const std::vector<std::optional<aslam::cameras::GridCalibrationTargetObservation>>& observations_camera_H,
+    const aslam::cameras::GridCalibrationTargetBase::Ptr& target) {
+  if (observations_camera_L.size() != observations_camera_H.size()) {
+    throw std::runtime_error("The number of observations for both cameras must be the same.");
+  }
+  // --- Get the initial guess for the transformation ---
+  // As the median of all pnp solutions
+  // Iterate throught shared observations between cameras
+  sm::kinematics::Transformation T_L;
+  sm::kinematics::Transformation T_H;
+  std::vector<sm::kinematics::Transformation> transformations;
+  for (size_t i = 0; i < observations_camera_L.size(); ++i) {
+    if (!observations_camera_L[i].has_value() || !observations_camera_H[i].has_value()) {
+      continue;  // Skip if either observation is not available
+    }
+
+    auto success = geometry_camera_L->estimateTransformation(observations_camera_L[i].value(), T_L);
+    if (!success) {
+      SM_ERROR_STREAM("Failed to estimate transformation for camera L at index " << i);
+      continue;
+    }
+    success = geometry_camera_H->estimateTransformation(observations_camera_H[i].value(), T_H);
+    if (!success) {
+      SM_ERROR_STREAM("Failed to estimate transformation for camera H at index " << i);
+      continue;
+    }
+
+    // Store the transformation from camera L to camera H
+    transformations.push_back(T_H.inverse() * T_L);
+  }
+
+  std::vector<Eigen::Vector3d> translations;
+  std::transform(transformations.begin(), transformations.end(), std::back_inserter(translations),
+                 [](const sm::kinematics::Transformation& t) {
+                   return t.t();
+                 });
+  auto median_translation = kalibr2::math::median(translations);
+
+  std::vector<Eigen::Vector3d> rotation_parameters;
+  std::transform(transformations.begin(), transformations.end(), std::back_inserter(rotation_parameters),
+                 [](const sm::kinematics::Transformation& t) {
+                   return sm::kinematics::RotationVector().rotationMatrixToParameters(t.C());
+                 });
+  auto median_rotation_parameters = kalibr2::math::median(rotation_parameters);
```

Comment:
@frneer meta: is this how the median rotation used to be computed? The whole concept of a median needs generalization for multi-dimensional settings (see https://en.wikipedia.org/wiki/Geometric_median) and I haven't worked out the math but I'm almost sure that element-wise median != geometric median.